internal/pkg/peer/blocksprovider/bft_censorship_monitor.go

```diff
@@ -0,0 +1,277 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package blocksprovider
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/internal/pkg/peer/orderers"
+	"github.com/pkg/errors"
+)
+
+type BlockProgressReporter interface {
+	BlockProgress() (uint64, time.Time)
+}
+
+// We ask for a stream of headers from all sources except the one supplying blocks.
+// We track the progress of header receivers against the block reception progress.
+// If there is a header that is ahead of the last block, and a timeout had passed since that header was received, we
+// declare that censorship was detected.
+// When censorship is detected, errCensorship is sent to the errorCh which can be read by ErrorsChannel() method.
+type bftCensorshipMonitor struct {
+	chainID          string
+	headerVerifier   BlockVerifier
+	requester        *DeliveryRequester
+	fetchSources     []*orderers.Endpoint
+	blockSourceIndex int
```

Comment:
the source index and sources are never updated.

Reply:
True, it is only set in the constructor method. The current plan is to stop the monitor and recreate it when ever we 
- refresh endpoints
- detect censorship (i.e. change the fetch index)

Since these are rare events, I prefer simplicity.